### PR TITLE
Issue 2585: (SegmentStore) Reduce excessive logging in some situations

### LINKED ISSUE #2585: Excessive logging in BookKeeperLog
**Problem description**
This line is logged way too many times. It shouldn't be.
`DEBUG i.p.s.s.i.bookkeeper.BookKeeperLog - Log[1]: Executing 0 writes.`

**Problem location**
BookKeeperLog.java: `processPendingWrites()`

**Suggestions for an improvement**
Don't log if nothing to write.

### PR DESCRIPTION
**Pull request by andreipaduroiu**
**Change log description**
- Fixed a condition in `BookKeeperLog.processPendingWrites` (Write Processor) where it would continuously invoke itself in cases when there was no work to be done.
    - Even if it doesn't call itself directly, it always invokes the Rollover Processor, which, in turn, always invokes the Write Processor.
    - This was fixed by not attempting to execute any writes or invoke the Rollover Processor if we currently have no writes to execute. 
- Fixed a condition in the Write Processor where it would log spurious, yet harmless `ObjectClosedException` or `NullPointerException` when closing the `BookKeeperLog` and there are still writes pending in the queue.

**Purpose of the change**
Fixes #2585.

**What the code does**
1. Not attempting to execute writes or invoke the Rollover Processor if there aren't any writes to execute.
2. Not executing the Write Processor if BookKeeperLog is closed.

**How to verify it**
Existing unit and integration tests must pass.

### CODE REVIEW

File: segmentstore/storage/impl/src/main/java/io/pravega/segmentstore/storage/impl/bookkeeper/BookKeeperLog.java [hunk 1]
```diff
@@ -400,37 +400,39 @@ private boolean processPendingWrites() {
...
-                w.getWriteLedger().ledger.asyncAddEntry(w.data.array(), w.data.arrayOffset(), w.data.getLength(), this::addCallback, w);
-            } catch (Throwable ex) {
-                // Synchronous failure (or RetriesExhausted). Fail current write.
-                boolean isFinal = !isRetryable(ex);
-                w.fail(ex, isFinal);
-
-                // And fail all remaining writes as well.
-                for (int j = i + 1; j < toExecute.size(); j++) {
-                    toExecute.get(j).fail(new DurableDataLogException("Previous write failed.", ex), isFinal);
+        if (toExecute.size() > 0) {
```
Comment (tkaitchuck) [inline segmentstore/storage/impl/src/main/java/io/pravega/segmentstore/storage/impl/bookkeeper/BookKeeperLog.java]: It might be simpler if this were inverted, so it could be handled as a special case where it just returned true.

Comment (andreipaduroiu) [reply segmentstore/storage/impl/src/main/java/io/pravega/segmentstore/storage/impl/bookkeeper/BookKeeperLog.java]: That's a good suggestion. It turns out if I do that I'll have too many `return` statements in this method so Checkstyle won't be happy.

I refactored the method and extracted some code out into two different methods.

File: segmentstore/storage/impl/src/main/java/io/pravega/segmentstore/storage/impl/bookkeeper/WriteQueue.java [hunk 2]
```diff
@@ -169,9 +168,11 @@ synchronized void add(Write write) {
      * Removes all the completed writes (whether successful or failed) from the beginning of the queue, until the first
      * non-completed item is encountered or the queue is empty.
      *
-     * @return True if there are items left in the queue, false otherwise.
+     * @return A CleanupStatus representing the state of the Operation. If there were failed writes, this will be WriteFailed,
+     * otherwise it will be one of QueueEmpty or QueueNotEmpty, depending on the final state of the queue when this method
+     * finishes.
      */
-    synchronized EnumSet<CleanupStatus> removeFinishedWrites() {
+    synchronized CleanupStatus removeFinishedWrites() {
```
Comment (fpj) [inline segmentstore/storage/impl/src/main/java/io/pravega/segmentstore/storage/impl/bookkeeper/WriteQueue.java]: The change to this class is returning a status rather than an enum set of statuses. I tried to understand from the PR description whether this is the change that is preventing the spurious exceptions. If this is the case, could you elaborate on how this is being achieved? Otherwise, what's the purpose of the change.

File: segmentstore/storage/impl/src/main/java/io/pravega/segmentstore/storage/impl/bookkeeper/BookKeeperLog.java [hunk 3]
```diff
@@ -354,6 +354,11 @@ public QueueStats getQueueStatistics() {
      * Write Processor main loop. This method is not thread safe and should only be invoked as part of the Write Processor.
      */
     private void processWritesSync() {
```
Comment (fpj) [inline segmentstore/storage/impl/src/main/java/io/pravega/segmentstore/storage/impl/bookkeeper/BookKeeperLog.java]: To confirm, even though there is a change in the internal behavior of this method, there is already enough test cases covering this code path? I'm trying to understand whether we need more testing.

File: segmentstore/storage/impl/src/main/java/io/pravega/segmentstore/storage/impl/bookkeeper/WriteQueue.java (hunk 2 above)
Comment (andreipaduroiu) [reply segmentstore/storage/impl/src/main/java/io/pravega/segmentstore/storage/impl/bookkeeper/WriteQueue.java]: 1. I tried to solve the problem by doing adding an early `return` statement from the method logging that message 
2. Checkstyle started complaining that there were too many `return` statements in that method.
3. I had to refactor the method to have fewer return statements.
4. During this I either had to make the code using this method a lot more uglier (another nested `if`) or refactor this method to return a single Enum (it really, only returned one useful value to begin with anyway) and keep the original method still look decent.

File: segmentstore/storage/impl/src/main/java/io/pravega/segmentstore/storage/impl/bookkeeper/BookKeeperLog.java (hunk 3 above)
Comment (andreipaduroiu) [reply segmentstore/storage/impl/src/main/java/io/pravega/segmentstore/storage/impl/bookkeeper/BookKeeperLog.java]: There are enough tests for this class.
